### ISSUE TITLE
Possible Json Serializer Multi Threading Issue 

### ISSUE THREAD

**Issue by gowthaman-basuvaraj** [svenkubiak/mangooio#299]
I have been noticing for some time that sometimes, (1 in 50 requests) there is json response data corruption.
I could never manage to figure out why so, it happens both in production and development mode.

i suspect that Json Serailizer is not thread safe?

Just managed to take screen shot when this happened

![image](https://cloud.githubusercontent.com/assets/117379/25192710/f2e1bbb0-2551-11e7-8372-53a7aff9bf9d.png)

![image](https://cloud.githubusercontent.com/assets/117379/25192723/018a2350-2552-11e7-920e-0f577062f3df.png)

you could see that 2 different requests ended up getting a same invalid json response

`[{"id":4,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":3.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":4,"createdAt":1492618774660,"modifiedAt":1492618774660,"createdBy":{"name":"Gowthaman Basuvaraj",{"name":"Gowthaman Basuvaraj","email":"gowthaman@365build.in","id":1,"phone":"9844816712","tenantAdmin":"entityStatus":"ACTIVE"},true,"superAdmin":"entityStatus":"ACT,VEEAT","enabled":true,"tenant":"companies":[{"id":1,1],"name":""offices","APPRO":,[1],"systemRoles":[1,2,3,4,5},,7,8,"company":9,{10,11,"id":12]1,,"id":},,"createdAt":1version36711rmission":,1,"modifiedAt"},1492618536711,{"date":"entityStatus":1490985000000"ACTIVEt","vendor":{"name":"Gowthaman Basuvaraj","tenant":{"id":1,"id":1},"company":{"entityStatus":"ACTIVE","id":1},"version":1}}]"docType":"PO","sentVia":"PHONE","documentNo":"PO-003","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":7,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":12.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":100.0}],"id":7,"createdAt":1492619922227,"modifiedAt":1492619922227,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"PO","sentVia":"PHONE","documentNo":"PO-004","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":9,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":1.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":9,"createdAt":1492620625319,"modifiedAt":1492620625319,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"QUOTATION","sentVia":"PHONE","documentNo":"Q-001","vendorReference":"VQ-001","office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":3,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":10.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":3,"createdAt":1492618719442,"modifiedAt":1492618719442,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"PO","sentVia":"PHONE","documentNo":"PO-002","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":6,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":10.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":6,"createdAt":1492619069630,"modifiedAt":1492619069630,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"PO","sentVia":"PHONE","documentNo":"PO-004","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":8,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":1.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":8,"createdAt":1492620242016,"modifiedAt":1492620242016,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"PO","sentVia":"PHONE","documentNo":"PO-002","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":5,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":10.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":300.0}],"id":5,"createdAt":1492618954551,"modifiedAt":1492618954551,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"vendor":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"docType":"PO","sentVia":"ORAL","documentNo":"PO-005","quotations":[3],"indents":[1,2],"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":10,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":5.0,"taxGroup":{"id":1,"entityStatus":"ACTIVE"},"basicPrice":500.0}],"id":10,"createdAt":1492620938241,"modifiedAt":1492620938241,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":1},{"date":1490985000000,"docType":"INDENT","sentVia":"PHONE","documentNo":"I-002","utilized":true,"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":2,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":5.0}],"id":2,"createdAt":1492618682333,"modifiedAt":1492620938270,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":3},{"date":1490985000000,"docType":"INDENT","sentVia":"PHONE","documentNo":"I-001","utilized":true,"office":{"name":"HO","id":1,"entityStatus":"ACTIVE"},"lineItems":[{"id":1,"slNo":1,"inventory":{"id":1,"entityStatus":"ACTIVE"},"qty":1.0}],"id":1,"createdAt":1492618654267,"modifiedAt":1492620938265,"createdBy":{"name":"Gowthaman Basuvaraj","id":1,"entityStatus":"ACTIVE"},"entityStatus":"ACTIVE","tenant":{"id":1,"name":"365Build"},"company":{"id":1},"version":3}]`

**Comment by svenkubiak** [comment]
Thanks for reporting! Will look into it.

**Comment by gowthaman-basuvaraj** [comment]
I tested with this sample program
`import com.google.common.collect.Lists;
import io.advantageous.boon.json.JsonSerializer;
import io.advantageous.boon.json.JsonSerializerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class MainClass {
    public static void main(String[] args) {
        List<Person> personList = new ArrayList<Person>();

        for (int i = 0; i < 1000; i++) {
            int max = new Random().nextInt(5);
            List<Address> addressList = new ArrayList<Address>();
            for (int j = 0; j < max; j++) {
                addressList.add(new Address("Addr " + j, 50000 + j));
            }

            personList.add(new Person("person-" + i, i, addressList));

        }

        final JsonSerializer json = new JsonSerializerFactory().create();
        Lists.partition(personList, 10)
                .parallelStream()
                .forEach(l -> {

                    String jsonStr = json.serialize(l).toString();

                    System.out.println(l.size() + " - " + jsonStr);
                });

    }

    public static class Person {
        String name;
        int age;
        List<Address> addresses;

        public Person(String name, int age, List<Address> addresses) {
            this.name = name;
            this.age = age;
            this.addresses = addresses;
        }
    }

    public static class Address {
        public String address;
        public int pin;

        public Address(String address, int pin) {
            this.address = address;
            this.pin = pin;
        }
    }
}
`

**Comment by svenkubiak** [comment]
Thanks for digging into this, but example code is probably not the best case to test. We need to check within the framework to find the issues. I have a suspicion but need to do some testing.

**Comment by svenkubiak** [comment]
Hm, first tests show not concurrency issues. This is the test case i used:

```
package io.mangoo.controllers.multithreaded;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.not;
import static org.hamcrest.Matchers.nullValue;

import java.util.UUID;

import org.junit.Test;

import io.mangoo.enums.ContentType;
import io.mangoo.test.utils.WebRequest;
import io.mangoo.test.utils.WebResponse;
import io.undertow.util.StatusCodes;

/**
 * 
 * @author svenkubiak
 *
 */
public class JsonMultithreadedTest {

    @Test
    public void test() throws InterruptedException {
        for (int p=0; p < 50; p++) {
            Runnable task = () -> {
                for (int j=0; j < 100; j++) {
                    //given
                    String uuid = UUID.randomUUID().toString();
                    String json = "{\"firstname\":\"$$\",\"lastname\":\"Parker\",\"age\":24}";
                    json = json.replace("$$", uuid);
                    
                    WebResponse response = WebRequest.post("/parse")
                            .withContentType(ContentType.APPLICATION_JSON)
                            .withRequestBody(json)
                            .execute();

                    //then
                    assertThat(response, not(nullValue()));
                    assertThat(response.getStatusCode(), equalTo(StatusCodes.OK));
                    assertThat(response.getContent(), equalTo(uuid + ";Parker;24"));    
                }
            };
            
            Thread thread = new Thread(task);
            thread.run();  
        }
    }
}
```

It spans 50 cuoncurrent threads which each execute a request 100 times. The request passes a newly generated UUID which is returned in the response after JSON is parsed.

**Comment by svenkubiak** [comment]
I have added a new ConcurrencyRunner for concurrency testing to the master branch.

**Comment by svenkubiak** [comment]
After some more testing it looks like converting an object to json is the problem. Probably due to the static utils method, when an mutable object is passed.

**Comment by svenkubiak** [comment]
Bug fix release 4.3.3. is out. I have also create a follow up issue to validate the utility-class in terms of thread-safety. 

Thanks again for reporting this! Please re-open if error still occurs.
